taskcat/taskcat-v9
# Stacker class

## Overview

Adds a class to handle groups of `Stack` instances needed for `taskcat test` commands. For this purpose the `Stacker` class was ~~rewritten~~ created. It takes a `Config` instance as an input and can either `create_stacks` or import `from_existing` (to support decoupled/stateless use of taskcat in the future). Should additional deployment mechanisms (not cloudformation) be added in the future, they would need to implement only the public API's of `Template` and `Stacker` classes and should work as drop in replacements to the cloudformation implementation.

Operations that involve use of cloudformation API's are grouped into concurrency groups to ensure the highest level of concurrency while minimizing the risk of throttling. `Stack` properties and calls to `events`/`resources`/`children` use a cache (default ttl 1min) to make repeated calls fast and reduce api load.

this pr also sneaks in a change to bucket creation behavior, enforcing a particular region for each AWS partition, this was needed to unblock functional testing of nested stacks, and bring it inline with the QuickStart bucket strategy. I've opened #42 to track making this more flexible in the future.

## Testing/Steps taken to ensure quality

manual testing for now, would like to get some feedback and a `taskcat test` implementation done before spending effort on unit tests for what's likely to see a fair amount of short-term churn.

## Notes

There are some minor changes to other parts of taskcat included in this pr that were needed for functional testing to work.

Also squeezed in changes to catch the use of print statements, which can be handy during debugging, but we want to use logging module for all output

## Testing Instructions

setup a config:
```python
from pathlib import Path
from taskcat import Config

c = Config(
    project_config_path=Path('./tests/data/nested_fail/ci/taskcat.yml'
                             '').resolve(),
    project_root=Path('./tests/data/nested_fail/').resolve()
)
```

create the stacks:

```python
from taskcat._cfn.threaded import Stacker
from taskcat._s3_stage import stage_in_s3

stage_in_s3(c)
s = Stacker(c)
s.create_stacks()
```

Check the status:

```python
s.status()

Out[1]: 
{'IN_PROGRESS': {},
 'COMPLETE': {},
 'FAILED': {'arn:aws:cloudformation:us-west-2:...': 'The following resource(s) failed to create: [ChildStack]. ',
  'arn:aws:cloudformation:us-east-1:....': 'The following resource(s) failed to create: [ChildStack]. '}}
```

hmm, both stacks failed, let's look at what/why:

```python
for resources in s.resources(status='CREATE_FAILED').values():
    for resource in resources:
        print(f'{stack_id} resource {resource.logical_id} of type {resource.type} failed')
        
arn:aws:cloudformation:us-east-1:.... resource ChildStack of type AWS::CloudFormation::Stack failed
arn:aws:cloudformation:us-west-2:... resource ChildStack of type AWS::CloudFormation::Stack failed
```

Notice the filter being applied with `s.resources(status='CREATE_FAILED')` any property of an `Stack`, `Event` or `Resource` can be used as a filter. If multiple filters are passed they work as AND conditions.

Testing nested stacks is a bit involved, as the magic params `$[tasckcat_bucket_name]` have not been wired in yet. So performing the tests below involves setting a breakpoint just before stack launch to set the value of `QSS3Bucket` to the generated bucket name.

```
from pathlib import Path
from taskcat import Config
from taskcat._cfn.threaded import Stacker
from taskcat._s3_stage import stage_in_s3
c = Config(
    project_config_path=Path('./tests/data/nested-fail/ci/taskcat.yml'
                             '').resolve(),
    project_root=Path('./tests/data/nested-fail/').resolve()
)
stage_in_s3(c)
s = Stacker(c)
s.create_stacks()
```

after a bit you can see the child stack:
```
s.stacks[0].children()
Out[7]: [<Stack object tCat-nested-fail-taskcat-json-1e2e315e06a94a9e8962b60347a228a7-ChildStack-CRSYHC6ZUDVX at 0x10bbf7eb8>]
```

finding children of children and their descendants can be done easily:
```
s.stacks[0].descendants()
Out[8]: 
[<Stack object tCat-nested-fail-taskcat-json-1e2e315e06a94a9e8962b60347a228a7-ChildStack-CRSYHC6ZUDVX at 0x10bbf7eb8>,
 <Stack object tCat-nested-fail-taskcat-json-1e2e315e06a94a9e8962b60347a228a7-ChildStack-Middle2Stack-L8E0QIBB0WI at 0x10c9fac18>,
 <Stack object tCat-nested-fail-taskcat-json-1e2e315e06a94a9e8962b60347a228a7-ChildStack-C-ChildStack-15DCBOEQ195T0 at 0x10ca74fd0>,
 <Stack object tCat-nested-fail-taskcat-json-1e2e315e06a94a9e8962b60347a228a7-ChildStack-M-ChildStack-19D1FG18JHOFK at 0x10ca74908>,
 <Stack object tCat-nested-fail-taskcat-json-1e2e315e06a94a9e8962b60347a228a7-ChildStack-M-ChildStack-10PR144QF0DDP at 0x10c3b9278>]
```

Once the stack fails:
```
s.stacks[0].status
Out[29]: 'CREATE_FAILED'
```
One can easily get the failure events without having to manually recurse through descendants:
```python
errors = s.stacks[0].error_events()
for error in errors:
    print(f"stack {error.stack_name} resource {error.logical_id} failed with error "
          f"{error.status_reason}")
```

outputs only the meaningful errors (not create cancelled or parent stack failures):
```
stack tCaT-nested-fail-taskcat-json-4600bbeff3d8479db842e1c79063dfd0-ChildStack-Y-ChildStack-K7TDJA8M86LP resource CustomResource failed with error Failed to create resource. See the details in CloudWatch Log Stream: 2019/08/19/[$LATEST]2ea497f9496740b29a6a8f9bc594b836
stack tCaT-nested-fail-taskcat-json-4600bbeff3d8479db842e1c79063dfd0-ChildStack-M-ChildStack-I6PBH6G7DGN4 resource CustomResource failed with error Failed to create resource. See the details in CloudWatch Log Stream: 2019/08/19/[$LATEST]cca9ad954e2e4f668685fb7ee692c65a
```
